core/arch/arm/plat-imx/imx-regs.h

```diff
@@ -249,4 +249,12 @@
 #define GPC_CPU_PGC_SW_PDN_REQ		0xfc
 #define GPC_PU_PGC_SW_PDN_REQ		0x104
 #define GPC_PGC_SW_PDN_PUP_REQ_CORE1_MASK BIT(1)
+
```

Comment:
Would you be able to put this  in the tzc380.c file instead?
this is a workaround for your choice of implementation and should be limited to the TzC380 file.

Also in the imx-regs.h file can you fix the TZASC2_BASE and only define it for the SoC that actually have a second TZ380 controller?
thanks

Reply:
No. Please coordinate with @clementfaure who wanted this define in the header file. `TZASC2_BASE` needs to be defined, otherwise the code won't compile. Dead code elimination will remove the code during compilation and a runtime assert ensures that `TZASC2_BASE` is set to a valid value which is non-zero.

Reply:
You are saying no to what in particular?
TZASC2_BASE need to be defined for SoC that actually have a second controller. 6UL and ULL (and all the other i.MX line) do not. i can understand that you dislike #ifdef in c code. But defining base address to a controller that does not exist is IMO horrible.

Now, for the TZASC2_BASE set to 0 is a consequence of your implementation choice. it is local to the tzc380.c file and does not have to be propagated to other c files. So imo it should not be imx-regs.h. 
when it comes to @clementfaure i think we are in sync. he was refering to the OFFSET for the Base address to the second controller that you have removed.

finally, for  what it is worth, i am still not sure what all this runtime check actually brings except replacing a ifdef CFG_XXX with a if(IS_ENABLED(CFG_XXX)

@MrVan , Would you comment on this? 

thanks
Silvano

Reply:
> You are saying no to what in particular?
> TZASC2_BASE need to be defined for SoC that actually have a second controller. 6UL and ULL (and all the other i.MX line) do not. i can understand that you dislike #ifdef in c code. But defining base address to a controller that does not exist is IMO horrible.

What does it matter for? Which other part of the operating system wants to access the `TZASC` and therefore needs valid base addresses? I cede that this could be moved to the `.c` file.

> finally, for what it is worth, i am still not sure what all this runtime check actually brings except replacing a ifdef CFG_XXX with a if(IS_ENABLED(CFG_XXX)

This is not a runtime check, as I have explained before. This will be evaluated to `0` or `1` by the preprocessor and is removed by the compilers dead code elimination pass if `0`.
The advantage being that all code is compiled, ensuring that there are no combinations which do not compile, which can be a problem with `#ifdef`s.

Reply:
> > You are saying no to what in particular?
> > TZASC2_BASE need to be defined for SoC that actually have a second controller. 6UL and ULL (and all the other i.MX line) do not. i can understand that you dislike #ifdef in c code. But defining base address to a controller that does not exist is IMO horrible.
> 
> What does it matter for? 

it does matter (at least to me) to have things rights. defining TZASC2_BASE for a SoC that does not have the controller is plain wrong. At best it is misleading the reader in thinking that all i.MX6 have a second TZASC controller, at worst someone actually use the define and makes the SoC Crash.i am amazed that i need to argue that much for something 

> Which other part of the operating system wants to access the `TZASC` and therefore needs > valid base addresses? 

The defined is in a global include file that is accessible by every C file. It must be correct regardless of it is use or not.

>I cede that this could be moved to the `.c` file.

please do it then.
> 
> > finally, for what it is worth, i am still not sure what all this runtime check actually brings except replacing a ifdef CFG_XXX with a if(IS_ENABLED(CFG_XXX)
> 
> This is not a runtime check, as I have explained before. This will be evaluated to `0` or `1` by the preprocessor and is removed by the compilers dead code elimination pass if `0`.
i understand that there are no extra line that will be executed as they will be optimized out, and for me that was beside the point as this is a configuration path and it is acceptatble to add few more cycle for those kind of test

> The advantage being that all code is compiled, ensuring that there are no combinations which do not compile, which can be a problem with `#ifdef`s.

It is ironic you use that argument for this addition, as your first patch submission actually broke the build and the second one workaround it by adding artificially a define to cope with the error (which is in the end what i do not like with your patch).

I hope you do realize that with the state of the i.MX build you need to compile one binary for each i.MX board supported, and that in the end, your change is not ensuring a lot more.

It would have been probably better to actually use a runtime check using soc_is_imxxxx defined in imx-common.c file and then gradualy try to remove the CFG_IMX6XX (or at least make it configuratble) like it is done in the linux kernel so that it would be possible to have one binary to address all (or at least more that one) i.MX6 devices

Reply:
>it does matter (at least to me) to have things rights. defining TZASC2_BASE for a SoC that does not have the controller is plain wrong. At best it is misleading the reader in thinking that all i.MX6 have a second TZASC controller, at worst someone actually use the define and makes the SoC Crash.i am amazed that i need to argue that much for something

Which does not answer my question. Which user expects the `TZASC2_BASE` to be a valid source of SoC information and would need it for the implementation? The only user of `TZASC2_BASE` I can think of is the `TZC380` driver which has been taken care of by this PR.

Talking about inconsistencies: `CAAM_BASE` is also defined for i.MX6ULL platforms. Why? Because its easier to let the define stay even for the i.MX6ULL by instead ensuring that the code which would use the define is not called at all. Which is the same thing I am doing in this PR.
Sure this is legacy code which has been there, but was also not fixed by the recent cleanups.
> I hope you do realize that with the state of the i.MX build you need to compile one binary for each i.MX board supported, and that in the end, your change is not ensuring a lot more.

Which is also the expected state by OP-TEE. OP-TEE does not have multi board support ATM and I don't know whether this is a project design goal.

> It would have been probably better to actually use a runtime check using soc_is_imxxxx defined in imx-common.c file and then gradualy try to remove the CFG_IMX6XX (or at least make it configuratble) like it is done in the linux kernel so that it would be possible to have one binary to address all (or at least more that one) i.MX6 devices

Now you are throwing me into a loop here, because this whole change from using the `soc_is_xxx` functions was actually suggested by the platform maintainer, discussed in the very first discussion in this PR.

Reply:
> OP-TEE does not have multi board support ATM and I don't know whether this is a project design goal.

Probably not short-term, but it would certainly be a nice thing to aim for. We already have introduced DT, it is a first step in that direction. And take for instance the STM32 platform, it uses (or will user) DT quite a lot, and probably aims at reducing the number of binaries. @etienne-lms correct me if I'm wrong ;-)

Reply:
> > it does matter (at least to me) to have things rights. defining TZASC2_BASE for a SoC that does not have the controller is plain wrong. At best it is misleading the reader in thinking that all i.MX6 have a second TZASC controller, at worst someone actually use the define and makes the SoC Crash.i am amazed that i need to argue that much for something
> 
> Which does not answer my question. Which user expects the `TZASC2_BASE` to be a valid source of SoC information and would need it for the implementation? The only user of `TZASC2_BASE` I can think of is the `TZC380` driver which has been taken care of by this PR.
> 

Really? I did not realize that i would have to argue that much for such simple change.

but let's continue that discussion and find an example. What about a debug facily that dump registers in some buffer for post mortem analysis.


> Talking about inconsistencies: `CAAM_BASE` is also defined for i.MX6ULL platforms. Why? Because its easier to let the define stay even for the i.MX6ULL by instead ensuring that the code which would use the define is not called at all. Which is the same thing I am doing in this PR.

I would not say it is easier but just laziness.  It is not because things are not clean that we should not spend time in the cleanup.
you proably can correct the CAAM inconsitency in your other pull request maybe :)

> Sure this is legacy code which has been there, but was also not fixed by the recent cleanups.

Again, cleanup is incremental, you could have started the trend.

> 
> > I hope you do realize that with the state of the i.MX build you need to compile one binary for each i.MX board supported, and that in the end, your change is not ensuring a lot more.
> 
> Which is also the expected state by OP-TEE. OP-TEE does not have multi board support ATM and I don't know whether this is a project design goal.
> 
> > It would have been probably better to actually use a runtime check using soc_is_imxxxx defined in imx-common.c file and then gradualy try to remove the CFG_IMX6XX (or at least make it configuratble) like it is done in the linux kernel so that it would be possible to have one binary to address all (or at least more that one) i.MX6 devices
> 
> Now you are throwing me into a loop here, because this whole change from using the `soc_is_xxx` functions was actually suggested by the platform maintainer, discussed in the very first discussion in this PR.

i really do not want to start yet another discussion so you can forget about this one.
i wanted just from you to make few cleanup in order for us to continue the upstream process of the code that is currently in the vendor repo. 
if you do not want to do it here we will do that as part of the #3169

Reply:
Let me get this straight:
We had this whole discussion because there is debugging code in your downstream vendor tree which checks the `TZASC2_BASE` and accesses it if its defined? I hate to break it to you but upstream changes may always break downstream patches. IMO this is not a sufficient reason to block this PR.

Regarding multi-board support: This whole discussion would be mood for multi-board support since the two controllers would be probed from the secure device tree anyway, however many controllers may be available. Is this an eventual goal of the NXP upstreaming efforts? If so are there documents or discussions on the standardization of the new bindings required for i.e. the TZC380 ?